src/main/java/com/amazon/opendistroforelasticsearch/security/transport/DefaultInterClusterRequestEvaluator.java

```diff
@@ -47,15 +56,34 @@
 import com.amazon.opendistroforelasticsearch.security.support.ConfigConstants;
 import com.amazon.opendistroforelasticsearch.security.support.WildcardMatcher;
 
-public final class DefaultInterClusterRequestEvaluator implements InterClusterRequestEvaluator {
+public final class DefaultInterClusterRequestEvaluator implements InterClusterRequestEvaluator, DCFListener {
 
     private final Logger log = LogManager.getLogger(this.getClass());
     private final String certOid;
-    private final List<String> nodesDn;
+    private final List<String> staticNodesDnFromEsYml;
+    private boolean dynamicNodesDnConfigEnabled;
+    private volatile Map<String, List<String>> dynamicNodesDn;
```

Comment:
We are already listening changes and updating it
do we still need to make it volatile?

Reply:
> We are already listening changes and updating it
> do we still need to make it volatile?

Yes, it is mostly to ensure the writes are quickly picked up in the subsequent reads. Otherwise further reads could end up reading from cache and miss the update.

Reply:
Consider using `AtomicReference` instead of `volatile`: `    private final AtomicReference<Map<String, List<String>>> dynamicNodesDn = new AtomicReference<>(Collections.emptyMap());`

Reply:
> Consider using `AtomicReference` instead of `volatile`: `private final AtomicReference<Map<String, List<String>>> dynamicNodesDn = new AtomicReference<>(Collections.emptyMap());`

Since I don't need CAS semantics, it seems like using volatile direclty is simpler. Do you see any other benefits of using AtomicReference in this specific scenario?

Reply:
>     1. Can you please make sure if nodes_dn API calls are being audit logged properly?
>     2. Under which event it is being audit logged? Do we need to create separate audit log events?

The current audit coverage covers [limited REST API access audit logging](https://github.com/opendistro-for-elasticsearch/security/blob/master/src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/AbstractApiAction.java#L368-L375)

Since the Admin access is checked after the section in the above link, they are not logged currently. However this behavior is consistent with other security API such as SSL certs APIs that were recently added. Since these are all admin API, do you think we can add them collectively?

(It looks like security index updates aren't logged, or that would have had the required info).

>     3. Can we verify 6.8 to 7.1 migration?
>     4. Can we test if security admin tools are working as expected ?

Good point. This was already being tested in SecurityAdminMigrationTests, but added a non-empty file to improve coverage.

```bash
$ pwd
~/code/native_es_stuff/fork/security/data/utest_n1_ffork_1_t17447483324660_migration

$ cat nodes_dn.yml
---
cluster1:
  nodes_dn:
  - "cn=popeye"

$ cat v7/nodes_dn.yml
---
_meta:
  type: "nodesdn"
  config_version: 2
cluster1:
  nodes_dn:
  - "cn=popeye"
  ```

>     5. Do we need any code change for certificate oid?

Nope. CertOid is fall back logic if there is no match for nodes_dn. It is singular value and hence makes sense only in YML.